modules/table/src/main/java/org/apache/ignite/internal/table/partition/HashPartitionManagerImpl.java

```diff
@@ -0,0 +1,128 @@
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.IntStream;
+import org.apache.ignite.internal.marshaller.MarshallerException;
+import org.apache.ignite.internal.marshaller.MarshallersProvider;
+import org.apache.ignite.internal.replicator.TablePartitionId;
+import org.apache.ignite.internal.schema.BinaryRowEx;
+import org.apache.ignite.internal.schema.SchemaRegistry;
+import org.apache.ignite.internal.schema.marshaller.TupleMarshallerException;
+import org.apache.ignite.internal.schema.marshaller.TupleMarshallerImpl;
+import org.apache.ignite.internal.schema.marshaller.reflection.KvMarshallerImpl;
+import org.apache.ignite.internal.schema.row.Row;
+import org.apache.ignite.internal.table.InternalTable;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.table.Tuple;
+import org.apache.ignite.table.mapper.Mapper;
+import org.apache.ignite.table.partition.HashPartition;
+import org.apache.ignite.table.partition.PartitionManager;
+
+/**
+ * Implementation of {@link PartitionManager} for tables with hash partitions.
+ */
+public class HashPartitionManagerImpl implements PartitionManager<HashPartition> {
+    private final InternalTable table;
+
+    private final SchemaRegistry schemaReg;
+
+    private final MarshallersProvider marshallers;
+
+    /**
+     * Constructor.
+     *
+     * @param table Internal table.
+     * @param schemaReg Schema registry.
+     * @param marshallers Marshallers.
+     */
+    public HashPartitionManagerImpl(
+            InternalTable table,
+            SchemaRegistry schemaReg,
+            MarshallersProvider marshallers
+    ) {
+        this.table = table;
+        this.schemaReg = schemaReg;
+        this.marshallers = marshallers;
+    }
+
+    @Override
+    public CompletableFuture<ClusterNode> primaryReplicaAsync(HashPartition partition) {
+        return table.partitionLocation(new TablePartitionId(table.tableId(), partition.partitionId()));
+    }
+
+    @Override
+    public CompletableFuture<Map<HashPartition, ClusterNode>> primaryReplicasAsync() {
+        HashPartition[] allPartitions = IntStream.range(0, table.partitions())
+                .mapToObj(HashPartition::new)
+                .toArray(HashPartition[]::new);
+
+        CompletableFuture<?>[] futures = new CompletableFuture<?>[allPartitions.length];
+        for (int i = 0; i < allPartitions.length; i++) {
+            HashPartition partition = allPartitions[i];
+            futures[i] = table.partitionLocation(new TablePartitionId(table.tableId(), partition.partitionId()));
+        }
+
+        return allOf(futures)
+                .thenApply(unused -> {
+                    Map<HashPartition, ClusterNode> result = new HashMap<>();
+                    for (int i = 0; i < allPartitions.length; i++) {
+                        result.put(allPartitions[i], (ClusterNode) futures[i].join());
+                    }
+                    return result;
+                });
```

Comment:
* Avoid streams
* Avoid unnecessary `allPartitions` array
* Initialize collections with capacity when possible

```suggestion
        int partitions = table.partitions();
        CompletableFuture<?>[] futures = new CompletableFuture<?>[partitions];
        
        for (int i = 0; i < partitions; i++) {
            futures[i] = table.partitionLocation(new TablePartitionId(table.tableId(), i));
        }
        
        return allOf(futures)
                .thenApply(unused -> {
                    Map<HashPartition, ClusterNode> result = new HashMap<>(partitions);
                    for (int i = 0; i < partitions; i++) {
                        result.put(new HashPartition(i), (ClusterNode) futures[i].join());
                    }
                    return result;
                });
```